pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyClientCnx.java

```diff
@@ -20,43 +20,93 @@
-                          AuthData clientAuthData, String clientAuthMethod, int protocolVersion) {
+                          Supplier<CompletableFuture<AuthData>> clientAuthDataSupplier,
+                          String clientAuthMethod, int protocolVersion, boolean forwardAuthorizationCredentials) {
         super(conf, eventLoopGroup);
         this.clientAuthRole = clientAuthRole;
-        this.clientAuthData = clientAuthData;
         this.clientAuthMethod = clientAuthMethod;
         this.protocolVersion = protocolVersion;
+        this.forwardAuthorizationCredentials = forwardAuthorizationCredentials;
+        this.clientAuthDataSupplier = clientAuthDataSupplier;
     }
 
     @Override
-    protected ByteBuf newConnectCommand() throws Exception {
-        if (log.isDebugEnabled()) {
-            log.debug("New Connection opened via ProxyClientCnx with params clientAuthRole = {},"
-                            + " clientAuthData = {}, clientAuthMethod = {}",
+    protected CompletableFuture<ByteBuf> newConnectCommand() throws Exception {
+        authenticationDataProvider = authentication.getAuthData(remoteHostName);
+        AuthData authData = authenticationDataProvider.authenticate(AuthData.INIT_AUTH_DATA);
+
+        return clientAuthDataSupplier.get().thenApply(clientAuthData -> {
+            if (log.isDebugEnabled()) {
+                log.debug("New Connection opened via ProxyClientCnx with params clientAuthRole = {},"
+                        + " clientAuthData = {}, clientAuthMethod = {}",
                     clientAuthRole, clientAuthData, clientAuthMethod);
+            }
+
+            return Commands.newConnect(authentication.getAuthMethodName(), authData, this.protocolVersion,
+                PulsarVersion.getVersion(), proxyToTargetBrokerAddress, clientAuthRole, clientAuthData,
+                clientAuthMethod);
+        });
+    }
+
+    @Override
+    protected void handleAuthChallenge(CommandAuthChallenge authChallenge) {
+        boolean isRefresh = Arrays.equals(
+            AuthData.REFRESH_AUTH_DATA_BYTES,
+            authChallenge.getChallenge().getAuthData()
+        );
+
+        if (!forwardAuthorizationCredentials || !isRefresh) {
+            super.handleAuthChallenge(authChallenge);
+            return;
         }
 
-        authenticationDataProvider = authentication.getAuthData(remoteHostName);
-        AuthData authData = authenticationDataProvider.authenticate(AuthData.INIT_AUTH_DATA);
-        return Commands.newConnect(authentication.getAuthMethodName(), authData, this.protocolVersion,
-            PulsarVersion.getVersion(), proxyToTargetBrokerAddress, clientAuthRole, clientAuthData,
-            clientAuthMethod);
+        try {
+            clientAuthDataSupplier.get()
+                .thenAccept(authData -> sendAuthResponse(authData, clientAuthMethod));
+        } catch (Exception e) {
```

Comment:
how is it possible to have an Exception here ?
in case of failure we should deal with the error and send a response or close the connection

Reply:
Agree. I will fix.

Reply:
I've already fixed it, make a review please.